### PR TITLE
a responsive component for creating new blog

### PR DESCRIPTION
**Pull request by epe-wtag**
## **Description**

**Background**
The createBlog component is being used for executing the POST method on the API to create a new blog. It is also covering the PUT method so that the blog update feature can also take place

**Goals**
  > To execute POST method on the API
  > To create a user-friendly interface for the user to write a blog.

**Technical Hints**
  > Use VueJS functionalities to efficiently create the feature

**Acceptance Criteria**
 > Proper follow-up of the given figma

### CODE REVIEW

File: src/components/CreateBlog.vue [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+  };
+
+  try {
+    const response = await fetch('https://66bc281924da2de7ff69786f.mockapi.io/Blog/', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(blogData),
+    });
```
Comment (shm-wtag) [inline src/components/CreateBlog.vue]: Can you create a composable for this fetch so every time you call it don't need to send the full api endpoint and headers

File: src/components/CreateBlog.vue [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  const blogData = {
+    category: category.value,
+    title: title.value,
+    description: description.value,
+    created_at: Math.floor(Date.now() / 1000), 
+    author: {
+      id: props.author.id,
+      name: props.author.name,
+      image: props.author.image
+    },
```
Comment (shm-wtag) [inline src/components/CreateBlog.vue]: what about 
```suggestion
    author: props.author,
```

File: src/components/CreateBlog.vue [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+  margin-bottom: 15px;
+  align-items: center;
+  justify-content: center;
+  text-align: center;
+}
+
+h2 {
+  text-align: center;
+  margin-bottom: 20px;
+}
```
Comment (shm-wtag) [inline src/components/CreateBlog.vue]: do not directly target tag name. Use class instead

File: src/views/ProfilePage.vue [hunk 4]
```diff
@@ -204,32 +245,41 @@ main {
     margin: 0 auto;
     margin-top: 30px;
 }
+
 .social-icon-div {
     padding: 8px 8px 8px 8px;
```
Comment (shm-wtag) [inline src/views/ProfilePage.vue]:
```suggestion
    padding: 8px;
```

File: src/components/BlogList.vue [hunk 5]
```diff
@@ -65,24 +31,23 @@ onMounted(() => {
...
-  }
-);
+function formatDate(timestamp: number) {
+  const date = new Date(timestamp * 1000);
+  return new Intl.DateTimeFormat('en-US', {
+    year: 'numeric',
+    month: 'long',
+    day: 'numeric',
+  }).format(date);
+}
```
Comment (hhu-wtag) [inline src/components/BlogList.vue:41]: Let's move this file a `utility.js` file. Export the function from that file and use it here. Thanks

File: src/components/BlogList.vue [hunk 6]
```diff
@@ -65,24 +31,23 @@ onMounted(() => {
...
-
 <template>
   <div class="blog-list">
     <div 
-    v-for="(article, index) in articles" 
-    :key="index" 
-    class="blog-item"
-    @click="navigateToBlog(article.id)" 
+      v-for="(article, index) in articles" 
+      :key="index"
```
Comment (hhu-wtag) [inline src/components/BlogList.vue:48]: We should always avoid `index` as keys in template loops. See [why](https://stackoverflow.com/questions/44531510/why-not-always-use-the-index-as-the-key-in-a-vue-js-for-loop#:~:text=Because%20arrays%20are%20mutable.,better%20than%20using%20an%20index.)

What you can do is you can use `uuidv4` library to generate unique key for each articles and then use the id as keys

File: src/components/BlogList.vue [hunk 7]
```diff
@@ -65,24 +31,23 @@ onMounted(() => {
...
   <div class="blog-list">
     <div 
-    v-for="(article, index) in articles" 
-    :key="index" 
-    class="blog-item"
-    @click="navigateToBlog(article.id)" 
+      v-for="(article, index) in articles" 
+      :key="index" 
+      class="blog-item"
+      @click="navigateToBlog(article.id)"
```
Comment (hhu-wtag) [inline src/components/BlogList.vue:50]:
```suggestion
      @click="onArticleSelect(article.id)" 
```

Following html element event names (onclick, onblur, onfocus) we should follow the same convention for our custom function names.

File: src/components/BlogList.vue [hunk 8]
```diff
@@ -130,7 +95,7 @@ watch(
 
 .blog-image {
   width: 100%;
-  height: auto;
+  height: 240px;
```
Comment (hhu-wtag) [inline src/components/BlogList.vue:98]: Please read this `maramari` in stackoverflow and determine wether you should use `rem` or `px`

https://stackoverflow.com/questions/11799236/should-i-use-px-or-rem-value-units-in-my-css

File: src/components/CreateBlog.vue [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+  const blogData = {
+    category: category.value,
+    title: title.value,
+    description: description.value,
+    created_at: Math.floor(Date.now() / 1000),
+    author: props.author,
+    image: image.value,
+  };
+
+  await post('Blog', blogData);
```
Comment (hhu-wtag) [inline src/components/CreateBlog.vue:31]: What is this function? Please give it a meaningful name to avoid confusion

File: src/components/CreateBlog.vue [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+  if (error.value) {
+    console.error('Failed to create blog:', error.value);
+  } else {
+    router.push({ name: 'home' });
+  }
```
Comment (hhu-wtag) [inline src/components/CreateBlog.vue:37]: This reactively broken code is very hard to follow. A function can always return a tuple. Hence we should follow better API design.

```javascript
// some-composable.vue

export async function createBlog(payload){
  let res = null
  let err = null

  try{
    res = await this.$api.post("/some-uri", payload)
  }catch(e){
    err = e
  }finally{
    return [res, err]
  }
}

// CreateBlog.vue

const [res, err] = await createBlogPost(payload)

if(err){
  // Notify user about the error.
  
  return
}

router.push({
  name: 'home'
})
```

**Review by hhu-wtag — CHANGES_REQUESTED** [review]

File: src/components/BlogList.vue (hunk 5 above)
Comment (epe-wtag) [reply src/components/BlogList.vue:41]: Hello bhaia, I've moved the function to a different file named utility.ts

**Review by shm-wtag — APPROVED** [review]
